### PR TITLE
Backport bugfixes, infrastructural changes and perf improvements from the polymorphism feature branch

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR backports a selection of bugfixes, infrastructural changes and perf improvements that were made as part of the polymorphism feature branch in #53882. I have layered the changes into separate commits that should ideally be reviewed independently:

1. https://github.com/dotnet/runtime/commit/a858d3b11d9ae6e2a1fed8198ed564fd4a8e82b7 fixes a series of bugs that were discovered while prototyping.
2. https://github.com/dotnet/runtime/commit/99529f71f9467675f277db33fb2c29f30ab420d5 adds debug assertions around `Push()` and `Pop()` operations that assist in early detection of stack-corrupting bugs.
3. https://github.com/dotnet/runtime/commit/4ba2497f0282f0ef25c278db92cff234930ed6e4 reworks and simplifies the `WriteStack` and `ReadStack` implementations to
    - Use arrays instead of lists for storing previous stackframes: previous frames can now be passed by reference where needed.
    - Ensures that the `_count` field always accurately reflects the current stack depth.
    - Removes the `StackFrame.Reset()` methods and instead clears `Current` by assigning it to a `default` instance. This eliminates a class of bugs caused by dirty frames being reused in new contexts.
4. https://github.com/dotnet/runtime/commit/cb594301e1fa57bd1d47f011392a00b67bc2a873 takes advantage of recently added JIT optimizations (e.g. optimizing `typeof(T).IsValueType` expressions #1157) in the serialization hot path methods.

The changes should make it easier to incorporate the polymorphism infrastructural changes in the future. Overall the changes result in a slight increase in serialization performance, on average between 4-8% faster in our benchmark suite.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ protected internal JsonConverter()
             // In the future, this will be check for !IsSealed (and excluding value types).
             CanBePolymorphic = TypeToConvert == JsonTypeInfo.ObjectType;
             IsValueType = TypeToConvert.IsValueType;
-            CanBeNull = !IsValueType || TypeToConvert.IsNullableOfT();
+            CanBeNull = default(T) is null;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:25]: I did verify this is faster for value types (same as for reference types).

Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:25]: It seems like other usages of `TypeToConvert.IsNullableOfT` and `CanBeNull` should be changed or removed.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 2]
```diff
@@ -476,6 +493,7 @@ internal bool TryWriteDataExtensionProperty(Utf8JsonWriter writer, T value, Json
 
             // Ignore the naming policy for extension data.
             state.Current.IgnoreDictionaryKeyPolicy = true;
+            state.Current.DeclaredJsonPropertyInfo = state.Current.JsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:496]: Was not setting `DeclaredJsonPropertyInfo` a bug?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.String.cs [hunk 3]
```diff
@@ -125,9 +125,6 @@ private static string WriteUsingMetadata<TValue>(in TValue value, JsonTypeInfo?
                 throw new ArgumentNullException(nameof(jsonTypeInfo));
             }
 
-            WriteStack state = default;
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.String.cs:128]: Thanks for removing this; I believe you mentioned this a while ago.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs [hunk 4]
```diff
@@ -59,25 +66,19 @@ internal struct ReadStack
...
-            {
-                _previous = new List<ReadStackFrame>();
-            }
-
-            if (_count > _previous.Count)
+            if (_stack is null)
             {
-                // Need to allocate a new array element.
-                _previous.Add(Current);
+                _stack = new ReadStackFrame[4];
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs:76]: 4 seems like a reasonable default depth. We any benchmark validations of that performed?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs [hunk 5]
```diff
@@ -16,15 +16,25 @@ namespace System.Text.Json
...
     {
         /// <summary>
-        /// The number of stack frames when the continuation started.
+        /// Exposes the stackframe that is currently active.
         /// </summary>
-        private int _continuationCount;
+        public WriteStackFrame Current;
+
+        /// <summary>
+        /// Buffer containing all frames in the stack. For performance it is only populated for serialization depths > 1.
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:24]: `For performance it is only populated for serialization depths > 1.`

It appears `EnsurePushCapacity()` always initialize the array?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs [hunk 6]
```diff
@@ -125,36 +124,33 @@ public void Push()
...
+
+                // check if we are done
+                if (_continuationCount == _count)
                 {
-                    _count++;
+                    _continuationCount = 0;
                 }
             }
+
+            Debug.Assert(PropertyPath() is not null);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:153]: Why was this added? PropertyPath would never be null, even if no path (it would be ".")

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs [hunk 7]
```diff
@@ -343,17 +325,17 @@ public string PropertyPath()
 
             for (int i = 0; i < count - 1; i++)
             {
-                AppendStackFrame(sb, _previous[i]);
+                AppendStackFrame(sb, ref _stack[i]);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:328]: Nice add of using "ref" variable support here

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs [hunk 8]
```diff
@@ -343,17 +325,17 @@ public string PropertyPath()
...
             if (_continuationCount == 0)
             {
-                AppendStackFrame(sb, Current);
+                AppendStackFrame(sb, ref Current);
             }
 
             return sb.ToString();
 
-            void AppendStackFrame(StringBuilder sb, in WriteStackFrame frame)
+            static void AppendStackFrame(StringBuilder sb, ref WriteStackFrame frame)
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:338]: Interesting. Was a defensive copy made prior?

**Review by steveharter — APPROVED** [review]
Thanks for the refactoring. LGTM with a couple questions.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs (hunk 5 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:24]: Correct, but that only happens for `_count > 1`. This is in line with current behavior.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:338]: Yep, I believe the compiler does this because `WriteStackFrame` is a mutable struct.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs (hunk 6 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:153]: It's not particularly meaningful, I ony added it to force execution of `PropertyPath()` on every run. My changes for IAsyncEnumerable actually broke this method but we had no test coverage for it. Perhaps it might be clearer if I change it to
```csharp
#if DEBUG
_ = PropertyPath();
#endif
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs:76]: This is reusing the default initial capacity for `List<T>`. It seems like a reasonable expected depth for most scenaria so I thought I'd leave it unchanged.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:496]: Correct, this is now caught by the asserts I have added around the `Push()` and `Pop()` methods.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:25]: It's still being used by the nullable converter factory and in a few reflection contexts where the generic type is not known statically.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:25]: I've decided to keep `CanBeNull` in the few places that are not part of the hot path, since it's probably easier to parse compared to `default(T) is null`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:341]: This is also true for most of the `Converter<T>.Write*` methods that accept an `in T value` parameter whenever `T` is a mutable struct. I was considering marking those `ref` as well, but it's probably overkill given it's meant to optimize for the occasional large mutable struct that gets serialized.
